### PR TITLE
port endpoints to generic etcd

### PR DESCRIPTION
**Pull request by kazegusuri**
- introduce endpointStrategy and call validation via it
- expose DELETE method for endpoints
  - results in integration test change
  - This is not used services like controller manager, but used by master internally.
- service registry handles endpoints via endpoint.Registry instead of direct access

### CODE REVIEW

**Comment by roberthbailey** [comment]
@davidopp @smarterclayton Can one of you nominate an appropriate reviewer for this change?

File: pkg/registry/endpoint/etcd/etcd.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+)
+
+// rest implements a RESTStorage for endpoints against etcd
+type REST struct {
+	*etcdgeneric.Etcd
+}
+
+// NewREST returns a RESTStorage object that will work against endpoints.
+func NewREST(h tools.EtcdHelper) *REST {
```
Comment (smarterclayton) [inline pkg/registry/endpoint/etcd/etcd.go]: Can you change the name here to NewREST to match #5758?

File: pkg/registry/endpoint/etcd/etcd_test.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+func validChangedEndpoints() *api.Endpoints {
+	endpoints := validNewEndpoints()
+	endpoints.ResourceVersion = "1"
+	endpoints.Endpoints = []api.Endpoint{{IP: "baz"}, {IP: "bar"}}
+	return endpoints
+}
+
+func TestCreate(t *testing.T) {
```
Comment (smarterclayton) [inline pkg/registry/endpoint/etcd/etcd_test.go]: Add a resttest.TestDelete test case as well (see pods).

**Comment by smarterclayton** [comment]
A few comments

**Comment by thockin** [comment]
I have a PR in flight that will conflict with this, so the sooner it merges the happier I will be :)

File: pkg/registry/endpoint/etcd/etcd.go (hunk 1 above)
Comment (kazegusuri) [inline pkg/registry/endpoint/etcd/etcd.go]: done

File: pkg/registry/endpoint/etcd/etcd_test.go (hunk 2 above)
Comment (kazegusuri) [inline pkg/registry/endpoint/etcd/etcd_test.go]: done

**Comment by kazegusuri** [comment]
Rebased to HEAD. Now running e2e test.

**Comment by smarterclayton** [comment]
LGTM

**Comment by kazegusuri** [comment]
Verified e2e test passed.

**Comment by thockin** [comment]
@smarterclayton OK to go with this?

**Comment by smarterclayton** [comment]
Needs a rebase

**Comment by kazegusuri** [comment]
Sorry for delay. Rebased again and verfied e2e.

**Comment by kazegusuri** [comment]
@smarterclayton PTAL. once this is merged, rebase #5441 to this.

**Pull request merged**